File: fern/snippets/tool-calling/versioning/typescript/rollback.ts
import { Composio } from "@composio/core";

// Rollback Strategy 1: Update SDK initialization
async function rollbackViaSdk(): Promise<Composio> {
    // Previous configuration (problematic)
    // const composio = new Composio({
    //     apiKey: "YOUR_API_KEY",
    //     toolkitVersions: {
    //         github: "20250116_00"  // New version causing issues
    //     }
    // });
    
    // Rollback to stable version
    const composio = new Composio({
        apiKey: "YOUR_API_KEY",
        toolkitVersions: {
            github: "20250110_00"  // Previous stable version
        }
    });
    
    console.log("Rolled back GitHub toolkit to version 20250110_00");
    return composio;
}

// Rollback Strategy 2: Update environment variables
async function rollbackViaEnv(): Promise<Composio> {
    // Set the stable version
    process.env.COMPOSIO_TOOLKIT_VERSION_GITHUB = "20250110_00";
    
    // Initialize SDK (will use environment variable)
    const composio = new Composio({
        apiKey: "YOUR_API_KEY"
    });
    
    console.log("Rolled back via environment variable");
    return composio;
}

// Rollback Strategy 3: Emergency skip version check
async function emergencyRollback(): Promise<any> {
    const composio = new Composio({
        apiKey: "YOUR_API_KEY"
    });
    
    // Skip version check temporarily (use with caution)
    const result = await composio.tools.execute({
        toolSlug: "GITHUB_CREATE_ISSUE",
        userId: "user-123",
        arguments: {
            repo: "my-repo",
            title: "Critical fix",
            body: "Emergency deployment"
        },
        dangerouslySkipVersionCheck: true
    });
    
    console.log("⚠️ Executed with version check skipped - fix version config ASAP");
    return result;
}

File: fern/snippets/tool-calling/versioning/python/rollback.py
from composio import Composio
import os

# Rollback Strategy 1: Update SDK initialization
def rollback_via_sdk():
    """Rollback to previous version by updating SDK configuration"""
    
    # Previous configuration (problematic)
    # composio = Composio(
    #     api_key="YOUR_API_KEY",
    #     toolkit_versions={
    #         "github": "20250116_00"  # New version causing issues
    #     }
    # )
    
    # Rollback to stable version
    composio = Composio(
        api_key="YOUR_API_KEY",
        toolkit_versions={
            "github": "20250110_00"  # Previous stable version
        }
    )
    
    print("Rolled back GitHub toolkit to version 20250110_00")
    return composio

# Rollback Strategy 2: Update environment variables
def rollback_via_env():
    """Rollback by updating environment variables"""
    
    # Set the stable version
    os.environ["COMPOSIO_TOOLKIT_VERSION_GITHUB"] = "20250110_00"
    
    # Initialize SDK (will use environment variable)
    composio = Composio(api_key="YOUR_API_KEY")
    
    print("Rolled back via environment variable")
    return composio

# Rollback Strategy 3: Emergency skip version check
def emergency_rollback():
    """Emergency rollback when version system itself has issues"""
    
    composio = Composio(api_key="YOUR_API_KEY")
    
    # Skip version check temporarily (use with caution)
    result = composio.tools.execute(
        "GITHUB_CREATE_ISSUE",
        {
            "user_id": "user-123",
            "arguments": {
                "repo": "my-repo",
                "title": "Critical fix",
                "body": "Emergency deployment"
            },
            "dangerously_skip_version_check": True
        }
    )
    
    print("⚠️ Executed with version check skipped - fix version config ASAP")
    return result